src/appleseed/renderer/modeling/object/rectobject.cpp

```diff
@@ -0,0 +1,295 @@
+#include "foundation/utility/containers/dictionary.h"
+#include "foundation/utility/job/iabortswitch.h"
+#include "foundation/utility/searchpaths.h"
+#include "foundation/utility/string.h"
+
+using namespace foundation;
+
+namespace renderer
+{
+
+//
+// RectObject class implementation.
+//
+
+namespace
+{
+    const char* Model = "rect_object";
+}
+
+RectObject::RectObject(
+    const char*            name,
+    const ParamArray&      params)
+  : ProceduralObject(name, params)
+{
+}
+
+void RectObject::release()
+{
+    delete this;
+}
+
+const char* RectObject::get_model() const
+{
+    return Model;
+}
+
+bool RectObject::on_frame_begin(
+    const Project&         project,
+    const BaseGroup*       parent,
+    OnFrameBeginRecorder&  recorder,
+    IAbortSwitch*          abort_switch)
+{
+    if (!ProceduralObject::on_frame_begin(project, parent, recorder, abort_switch))
+        return false;
+
+    const double width  = get_uncached_width();
+    const double height = get_uncached_height();
+
+    m_half_width = width * 0.5;
+    m_half_height = height * 0.5;
+
+    if (width != 0.0)
+        m_rcp_width = 1.0 / width;
+
+    if (height != 0.0)
+        m_rcp_height = 1.0 / height;
+
+    return true;
+}
+
+GAABB3 RectObject::compute_local_bbox() const
+{
+    const float width  = static_cast<float>(get_uncached_width());
+    const float height = static_cast<float>(get_uncached_height());
+
+    const GVector3 pmax(width * 0.5f, 0.0f, height * 0.5f);
+    const GVector3 pmin(-pmax.x, 0.0f, -pmax.z);
+
+    GAABB3 bbox(pmin, pmax);
+    return bbox;
+}
+
+size_t RectObject::get_material_slot_count() const
+{
+    return 1;
+}
+
+const char* RectObject::get_material_slot(const size_t index) const
+{
+    return "default";
+}
+
+double RectObject::get_uncached_width() const
+{
+    return m_params.get_optional<double>("width", 1.0);
+}
+
+double RectObject::get_uncached_height() const
+{
+    return m_params.get_optional<double>("height", 1.0);
+}
+
+void RectObject::intersect(
+    const ShadingRay&      ray,
+    IntersectionResult&    result) const
+{
+    if (m_half_width == 0.0 || m_half_height == 0.0)
+    {
+        result.m_hit = false;
+        return;
+    }
+
+    if (ray.m_dir.y == 0.0)
+    {
+        result.m_hit = false;
+        return;
+    }
+
+    const double t = -ray.m_org.y / ray.m_dir.y;
+
+    if (t < ray.m_tmin || t >= ray.m_tmax)
+    {
+        result.m_hit = false;
+        return;
+    }
+
+    const double u =
+        ((ray.m_org.x + t * ray.m_dir.x) + m_half_width) * m_rcp_width;
+
+    if (u < 0.0 || u > 1.0)
+    {
+        result.m_hit = false;
+        return;
+    }
+
+    const double v =
+        ((ray.m_org.z + t * ray.m_dir.z) + m_half_height) * m_rcp_height;
+
+    if (v < 0.0 || v > 1.0)
+    {
+        result.m_hit = false;
+        return;
+    }
+
+    result.m_hit = true;
+    result.m_distance = t;
+
+    const Vector3d n(0.0, 1.0, 0.0);
```

Comment:
The default orientation of disks and rects should be the same. I don't really have a preference.